From 7ab7cb479fa229d59c7af63a1870916938cbd956 Mon Sep 17 00:00:00 2001
From: Sanyog Kale <skale@habana.ai>
Date: Wed, 10 Nov 2021 07:40:22 +0200
Subject: [PATCH] [SW-0] gaudi2: arc: update sleep routine of arc0

As part of secureboot flow, ARC0 is used as secondary CPU to
perform pcie initialization and sleep after initialization is complete.
Current logic doesn't put ARC0 into sleep due to which preboot
doesn't detect pcie link up. This patch update code to put ARC0
in sleep state once pcie initialization is done. As part of FIT
load process ARC0 is reset and then is used for running zephyr PID.

Change-Id: I96840fdb2a6ebddf30f3b760c82372ce3961f927
---

diff --git a/arch/gaudi2/arc/common.c b/arch/gaudi2/arc/common.c
index 351e35a..6a27d2f 100644
--- a/arch/gaudi2/arc/common.c
+++ b/arch/gaudi2/arc/common.c
@@ -291,8 +291,7 @@
 	preboot_pcie_monitoring();
 
 	/* Put secondary CPU in sleep state until further notice */
-	while (1)
-		wfe();
+	sleep();
 }
 
 void preboot_pcie_wait_for_link(void)
diff --git a/arch/gaudi2/arc/common.h b/arch/gaudi2/arc/common.h
index 5163eed..c337dc5 100644
--- a/arch/gaudi2/arc/common.h
+++ b/arch/gaudi2/arc/common.h
@@ -97,6 +97,8 @@
 
 #define ARC_ADDR_MASK		0x7FFFFFFull
 
+#define sleep()	__asm__ __volatile__("sleep\n\t" : : )
+
 #ifndef __ASSEMBLY__
 void init_secondary(void);
 extern void secondary_entry(void);
